src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -315,37 +319,49 @@ Status CoreWorkerDirectTaskSubmitter::KillTask(TaskSpecification task_spec,
 
           if (scheduled_tasks->second.empty()) {
             task_queues_.erase(scheduling_key);
+            CancelWorkerLeaseIfNeeded(scheduling_key);
           }
-          // Try to cancel lease requests
+          // Try to cancel the lease requests.
           task_finisher_->PendingTaskFailed(task_spec.TaskId(),
                                             rpc::ErrorType::TASK_CANCELLED);
           return Status::OK();
         }
       }
     }
-
-    cancelled_tasks_.emplace(task_spec.TaskId());
-    auto rpc_client = sent_tasks_.find(task_spec.TaskId());
-    // Checks for RPC handle for worker
-    if (rpc_client != sent_tasks_.end() &&
+    // This will get removed either when the RPC call to cancel is returned
+    // or when all dependencies are resolved.
+    RAY_CHECK(cancelled_tasks_.emplace(task_spec.TaskId()).second);
+    auto rpc_client = executing_tasks_.find(task_spec.TaskId());
+    // Looks for an RPC handle for the worker executing the task.
+    if (rpc_client != executing_tasks_.end() &&
         client_cache_.find(rpc_client->second) != client_cache_.end()) {
       client = client_cache_.find(rpc_client->second)->second;
     }
   }
 
+  // This case is reached for tasks that have unresolved dependencies.
   if (client == nullptr) {
     return Status::OK();
   }
 
-  auto request = rpc::KillTaskRequest();
+  auto request = rpc::CancelTaskRequest();
   request.set_intended_task_id(task_spec.TaskId().Binary());
   request.set_force_kill(force_kill);
-  RAY_UNUSED(client->KillTask(
-      request, [this, task_spec, force_kill, num_tries](const Status &status,
-                                                        const rpc::KillTaskReply &reply) {
-        if (status.ok() && !reply.attempt_succeeded() && num_tries > 0) {
-          usleep(RayConfig::instance().cancellation_retry_ms() * 1000);
-          RAY_UNUSED(KillTask(task_spec, force_kill, num_tries - 1));
+  RAY_UNUSED(client->CancelTask(
+      request, [this, task_spec, force_kill](const Status &status,
+                                             const rpc::CancelTaskReply &reply) {
+        absl::MutexLock lock(&mu_);
+        cancelled_tasks_.erase(task_spec.TaskId());
+        if (!reply.attempt_succeeded()) {
```

Comment:
It's probably good to check whether `status.OK()` also, since I'm not sure what the semantics are for `reply`'s value if there was a nerror.

Reply:
ok!